### PR TITLE
C#: Encode GodotProjectDir as Base64 to prevent issues with special characters

### LINKED ISSUE #73932: can_instantiate: Cannot instance script because the associated class could not be found (Project name contains invalid characters like `#`, `;`, `'`)
### Godot version

4.0RC5 Mono

### System information

Windows 10

### Issue description

I've seen this issue posted before and closed despite never being fixed

### Steps to reproduce

- create project
- create new scene with CharacterBody2D, let's call it "Player"
- add AnimatedSprite2D and CollosionObject2D to it as children
- create another scene and set it as main scene
- create new script that inherits from CharacterBody2D and attach by drag and drop to Player
- drag and drop Player scene to main scene
- run

### Minimal reproduction project

[Game2.zip](https://github.com/godotengine/godot/files/10831734/Game2.zip)

### PR DESCRIPTION
**Pull request by raulsntos**
- Fixes https://github.com/godotengine/godot/issues/73932.

### CODE REVIEW

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPathAttributeGenerator.cs [hunk 1]
```diff
@@ -22,12 +22,15 @@ public void Execute(GeneratorExecutionContext context)
 
             // NOTE: NotNullWhen diagnostics don't work on projects targeting .NET Standard 2.0
             // ReSharper disable once ReplaceWithStringIsNullOrEmpty
-            if (!context.TryGetGlobalAnalyzerProperty("GodotProjectDir", out string? godotProjectDir)
+            if (!context.TryGetGlobalAnalyzerProperty("GodotProjectDirBase64", out string? godotProjectDir)
                 || godotProjectDir!.Length == 0)
             {
                 throw new InvalidOperationException("Property 'GodotProjectDir' is null or empty.");
```
Comment (RedworkDE) [inline modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPathAttributeGenerator.cs:28]:
```suggestion
                throw new InvalidOperationException("Property 'GodotProjectDirBase64' is null or empty.");
```

Alternatively fallback to the old property for compatibility with existing projects that directly referencing the generator,

**Review by RedworkDE — CHANGES_REQUESTED** [review]

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPathAttributeGenerator.cs (hunk 1 above)
Comment (neikeq) [reply modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPathAttributeGenerator.cs:28]: Yes, even though `Godot.NET.Sdk` references the newer version, we should avoid breaking compat.

**Review by neikeq — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by geowarin** [comment]
After this PR my build fails:

`./modules/mono/build_scripts/build_assemblies.py --godot-output-dir=./bin \
  --push-nupkgs-local ~/dev/LocalNugetSource`

```
/home/geo/dev/godot/modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators.Sample/Godot.SourceGenerators.Sample.csproj(10,5): 
error MSB4186: Invalid static method invocation syntax: "[MSBuild]::ConvertToBase64('$(GodotProjectDir)')". 
Method '[MSBuild]::ConvertToBase64' not found. Static method invocation should be of the form: $([FullTypeName]::Method()), e.g. $([System.IO.Path]::Combine(`a`, `b`)). 
Check that all parameters are defined, are of the correct type, and are specified in the right order.
```

`dotnet --info`:

<details>

```
.NET SDK (reflecting any global.json):
 Version:   6.0.114
 Commit:    346c0065dd

Runtime Environment:
 OS Name:     arch
 OS Version:  
 OS Platform: Linux
 RID:         arch-x64
 Base Path:   /usr/share/dotnet/sdk/6.0.114/

Host:
  Version:      7.0.3
  Architecture: x64
  Commit:       0a2bda10e8

.NET SDKs installed:
  6.0.114 [/usr/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.NETCore.App 6.0.14 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

Other architectures found:
  None

Environment variables:
  DOTNET_ROOT       [/usr/share/dotnet]

global.json file:
  Not found

```

</details>

**Comment by raulsntos** [comment]
Your version of MSBuild is likely outdated, what is the output of `dotnet msbuild --version`? I think this was added in 17.3.0.

**Comment by geowarin** [comment]
Indeed, I have a version < 17.3.0.

```
❯ dotnet msbuild --version
Microsoft (R) Build Engine version 17.0.1+b177f8fa7 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

17.0.1.11501%   
```

How to I update it? I think it's bundled by my distro.

**Comment by raulsntos** [comment]
MSBuild is part of the .NET SDK, can you update it?

**Comment by geowarin** [comment]
I've flag the arch packages as out-of-date.

**Comment by neikeq** [comment]
I think we should look for an alternative that doesn't require a newer MSBuild. Otherwise, we can't include it in 4.0.x as it's a breaking change. Is there a way in MSBuild to check if a method is available?

**Comment by geowarin** [comment]
I agree with @neikeq project build are going to fail with this kind of errors:

image

Which is not user friendly at all.
And is a breaking change.

**Comment by YuriSizov** [comment]
Cherry-picked for 4.0.2 (alongside https://github.com/godotengine/godot/pull/74479).

**Pull request merged**